### PR TITLE
(Re)introduce ViewTreeLifecycleOwner support to containers 

### PR DESCRIPTION
**Pull request by zach-klippenstein**
Introduces something called `WorkflowLifecycleOwner` that implements `LifecycleOwner` and is set as the `ViewTreeLifecycleOwner` on the child views of the navigation containers: `WorkflowViewStub`, `BackstackContainer`, and `ModalContainer`. These containers all move their lifecycle to the `DESTROYED` state when they are about to discard the current child view and replace it with a new one because the rendering type changed. Until then, their lifecycle follows that of their parent lifecycle. 

This PR was created in March, then left for a few months, and finally picked back up in August. That said, very little had to be changed. The updates to AndroidX and Kotlin were merged in separate PRs (#459, #460). It was started by reverting commit 8500b0670e52d02a7904a91dda7914b57a62b968 (which reverted 74e14e9c and cdca31ce). 

TODO:
- [ ] Test with POS before merging. 
- [x] Update commit message to match new PR title & description.

### CODE REVIEW

**Review by rjrjr — APPROVED** [review]

**Review by laurakelly — APPROVED** [review]
Nice 🙏 🚀

**Comment by zach-klippenstein** [comment]
Waiting to merge this until I can verify it doesn't break POS.

**Review by rjrjr — APPROVED** [review]

**Comment by SalvatoreT** [comment]
🎉

**Comment by zach-klippenstein** [comment]
One signficant late-breaking change: I realized we can use `initializeView` to pull the `WorkflowLifecycleOwner` installation out of the `ViewRegistry` altogether and make it the responsibility of whatever parent is actually requesting the new view. This is cleaner because those views already needed to be aware of the WLO and call destroy on it at the right time, so this is much more symmetric. It also means that containers that don't care about lifecycle (such as the decorator) won't have a WLO unnecessarily set on them. Lastly, because `installOn` is now called by the containers themselves, it's much easier for `BackstackContainer` to pass its lifecycle ratchet as described [here](https://github.com/square/workflow-kotlin/issues/458).

File: workflow-ui/modal-android/src/main/java/com/squareup/workflow1/ui/modal/ModalContainer.kt [hunk 1]
```diff
@@ -42,6 +42,8 @@ public abstract class ModalContainer<ModalRenderingT : Any> @JvmOverloads constr
 
   private var dialogs: List<DialogRef<ModalRenderingT>> = emptyList()
 
+  private val parentLifecycleOwner by lazy(NONE) { WorkflowLifecycleOwner.get(this) }
```
Comment (steve-the-edwards) [inline workflow-ui/modal-android/src/main/java/com/squareup/workflow1/ui/modal/ModalContainer.kt:45]:
```suggestion
  private val parentLifecycleOwner by lazy(mode = LazyThreadSafetyMode.NONE) { WorkflowLifecycleOwner.get(this) }
```

Maybe just clearer?

**Review by steve-the-edwards — APPROVED** [review]

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowLifecycleOwner.kt [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+  private val localLifecycle =
+    if (enforceMainThread) LifecycleRegistry(this) else createUnsafe(this)
+
+  /**
+   * The parent lifecycle found by calling [ViewTreeLifecycleOwner.get] on the owning view's parent
+   * (once it's attached), or if no [ViewTreeLifecycleOwner] is set, then by trying to find a
+   * [LifecycleOwner] on the view's context.
+   *
+   * When the view is detached, we keep the reference to, and observing, the previous parent
```
Comment (rjrjr) [inline workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowLifecycleOwner.kt:118]:
```suggestion
   * When the view is detached, we keep the reference to, and observe, the previous parent
```

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowViewStub.kt [hunk 3]
```diff
@@ -203,12 +203,29 @@ public class WorkflowViewStub @JvmOverloads constructor(
     val parent = actual.parent as? ViewGroup
       ?: throw IllegalStateException("WorkflowViewStub must have a non-null ViewGroup parent")
 
+    // If we have a child view (i.e. this !== actual), then the old child view is going to
```
Comment (rjrjr) [inline workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowViewStub.kt:206]:
```suggestion
    // If we have a delegate view (i.e. this !== actual), then the old delegate is going to
```

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowViewStub.kt [hunk 4]
```diff
@@ -203,12 +203,29 @@ public class WorkflowViewStub @JvmOverloads constructor(
...
+    // eventually be detached by replaceOldViewInParent. When that happens, it's not just a regular
+    // detach, it's a navigation event that effectively says that view will never come back. Thus,
+    // we want its Lifecycle to move to permanently destroyed, even though the parent lifecycle is
+    // still probably alive.
+    // If actual === this, then this stub hasn't been initialized with a real child view yet. If
```
Comment (rjrjr) [inline workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowViewStub.kt:211]:
```suggestion
    //
    // If actual === this, then this stub hasn't been initialized with a real delegate view yet. If
```

**Review by rjrjr — APPROVED** [review]

File: workflow-ui/backstack-android/src/main/java/com/squareup/workflow1/ui/backstack/ViewStateCache.kt [hunk 5]
```diff
@@ -67,25 +68,28 @@ internal constructor(
...
-            viewStates += savedKey to ViewStateFrame(savedKey, saved)
+        ?.let { savedKey ->
+          val saved = SparseArray<Parcelable>().apply {
+            oldViewMaybe.saveHierarchyState(this)
           }
+          viewStates += savedKey to ViewStateFrame(savedKey, saved)
+        }
+
+      // Notify the view we're about to replace that it's going away.
+      WorkflowLifecycleOwner.get(oldViewMaybe)?.destroyOnDetach()
```
[the remark this answers is not in the document]
Comment (zach-klippenstein) [reply workflow-ui/backstack-android/src/main/java/com/squareup/workflow1/ui/backstack/ViewStateCache.kt:92]: `ViewStateCache` is _not_ duplicated by our internal fork, right?

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/ViewRegistry.kt [hunk 6]
```diff
@@ -109,6 +109,13 @@ public fun ViewRegistry(): ViewRegistry = TypedViewRegistry()
  * [AndroidViewRendering.viewFactory], if there is one. Note that this means that a
  * compile time [AndroidViewRendering.viewFactory] binding can be overridden at runtime.
  *
+ * The returned view will have a [WorkflowLifecycleOwner] set on it. The returned view msut EITHER:
```
[the remark this answers is not in the document]
Comment (zach-klippenstein) [reply workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/ViewRegistry.kt:112]: I thought i fixed this.

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowLifecycleOwner.kt [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+ * of the next-nearest [ViewTreeLifecycleOwner] above it (it mirrors its parent's lifecycle until
+ * it's destroyed), and can be [asked to destroy][destroyOnDetach] itself early.
+ *
+ * This type is meant to help integrate with [ViewTreeLifecycleOwner] by allowing the creation of a
+ * tree of [LifecycleOwner]s that mirrors the view tree.
+ *
+ * Custom container views that use [ViewRegistry.buildView] to create their children _must_ ensure
+ * they call [destroyOnDetach] on the outgoing view before they replace children with new views.
+ * If this is not done, then certain processes that are started by that view's subtree may continue
+ * to run long after the view is been detached, and memory and other resources may be leaked.
```
Comment (zach-klippenstein) [inline workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowLifecycleOwner.kt:32]:
```suggestion
 * to run long after the view has been detached, and memory and other resources may be leaked.
```

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowLifecycleOwner.kt (hunk 2 above)
Comment (zach-klippenstein) [reply workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowLifecycleOwner.kt:119]: The parent was already observed. I wanted this to say "we keep observing the previous parent."

File: samples/tictactoe/app/src/androidTest/java/com/squareup/sample/TicTacToeEspressoTest.kt [hunk 8]
```diff
@@ -38,25 +38,23 @@ import java.util.concurrent.atomic.AtomicReference
...
   fun setUp() {
-    scenario = launch(TicTacToeActivity::class.java)
-        .apply {
-          onActivity { activity ->
-            IdlingRegistry.getInstance()
-                .register(activity.idlingResource)
-            activity.requestedOrientation = SCREEN_ORIENTATION_PORTRAIT
-          }
-        }
+    scenario.onActivity { activity ->
```
[the remark this answers is not in the document]
Comment (zach-klippenstein) [reply samples/tictactoe/app/src/androidTest/java/com/squareup/sample/TicTacToeEspressoTest.kt:46]: I don't think so, we just didn't realize this API existed before.

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowAndroidXSupport.kt [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+ */
+public object WorkflowAndroidXSupport {
+
+  /**
+   * Tries to get the parent lifecycle from the current view via [ViewTreeLifecycleOwner], if that
+   * fails it looks up the context chain for a [LifecycleOwner], and if that fails it just returns
+   * null.
+   */
+  @WorkflowUiExperimentalApi
+  public fun lifecycleOwnerFromViewTreeOrContext(view: View): LifecycleOwner? =
```
[the remark this answers is not in the document]
Comment (zach-klippenstein) [reply workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowAndroidXSupport.kt:20]: I did: "if that fails it looks up the context chain for a [LifecycleOwner]", but I can add more text specifically comparing it with that.

- [ ] Clarify `lifecycleOwnerFromViewTreeOrContext`.

File: workflow-ui/backstack-android/src/main/java/com/squareup/workflow1/ui/backstack/ViewStateCache.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (zach-klippenstein) [reply workflow-ui/backstack-android/src/main/java/com/squareup/workflow1/ui/backstack/ViewStateCache.kt:92]: Actually we don't even need a hook, just move this line to after the `update` call in `BackstackContainer.update`.

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowAndroidXSupport.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (zach-klippenstein) [reply workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowAndroidXSupport.kt:20]: Oops, I totally had the wrong thing in mind. Clearly this needs more clarification.

File: workflow-ui/backstack-android/src/main/java/com/squareup/workflow1/ui/backstack/ViewStateCache.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (rjrjr) [reply workflow-ui/backstack-android/src/main/java/com/squareup/workflow1/ui/backstack/ViewStateCache.kt]: > `ViewStateCache` is not duplicated by our internal fork, right?

Correct.

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowLifecycleOwner.kt [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+  private val localLifecycle =
+    if (enforceMainThread) LifecycleRegistry(this) else createUnsafe(this)
+
+  /**
+   * The parent lifecycle found by calling [ViewTreeLifecycleOwner.get] on the owning view's parent
+   * (once it's attached), or if no [ViewTreeLifecycleOwner] is set, then by trying to find a
+   * [LifecycleOwner] on the view's context.
+   *
+   * When the view is detached, we keep the reference to, and observing, the previous parent
+   * lifecycle, to ensure we get destroyed correctly if the parent is destroyed while we're
```
Comment (rjrjr) [inline workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowLifecycleOwner.kt:120]:
```suggestion
   * When the view is detached, we keep the reference to the previous parent
   * lifecycle, and keep observing it, to ensure we get destroyed correctly if the 
   * parent is destroyed while we're
```

**Comment by zach-klippenstein** [comment]
Internal build system seems determined to reject snapshot builds. I think I'm gonna merge this and do a release.

**Pull request merged**